### PR TITLE
State refactoring

### PR DESCRIPTION
**Pull request by teseoch**
- [x] remove state references from everywhere
- [x] remove output stuff from state

### CODE REVIEW

File: src/polyfem/solver/forms/FrictionForm.hpp [hunk 1]
```diff
@@ -28,7 +26,9 @@ namespace polyfem::solver
 		/// @param dt Time step size
 		/// @param contact_form Pointer to contact form; necessary to have the barrier stiffnes, maybe clean me
 		FrictionForm(
-			const State &state,
+			const ipc::CollisionMesh &collision_mesh,
+			const Eigen::MatrixXd &boundary_nodes_pos,
+			const int dim,
```
Comment (zfergus) [inline src/polyfem/solver/forms/FrictionForm.hpp:31]: Could it get `dim` from `boundary_nodes_pos.cols()`?

File: src/polyfem/problem/KernelProblem.hpp [hunk 2]
```diff
@@ -32,7 +30,7 @@ namespace polyfem
 			void set_parameters(const json &params) override;
 			bool is_scalar() const override;
 
-			State *state;
+			assembler::AssemblerUtils *assembler;
```
Comment (zfergus) [inline src/polyfem/problem/KernelProblem.hpp:33]: Smart pointer?

File: src/polyfem/solver/forms/ContactForm.hpp [hunk 3]
```diff
@@ -22,8 +22,11 @@ namespace polyfem::solver
 		/// @param broad_phase_method Broad phase method to use for distance and CCD evaluations
 		/// @param ccd_tolerance Continuous collision detection tolerance
 		/// @param ccd_max_iterations Continuous collision detection maximum iterations
-		ContactForm(const State &state,
+		ContactForm(const ipc::CollisionMesh &collision_mesh,
+					const Eigen::MatrixXd &boundary_nodes_pos,
+					const int dim,
```
Comment (zfergus) [inline src/polyfem/solver/forms/ContactForm.hpp:27]: Same here, can we get dim from `boundary_node_pos.cols()`?

**Review by zfergus — APPROVED** [review]

**Comment by teseoch** [comment]
If this looks ok, we only need to update the doxygen

File: src/polyfem/CMakeLists.txt [hunk 4]
```diff
@@ -23,3 +23,4 @@ add_subdirectory(quadrature)
 add_subdirectory(solver)
 add_subdirectory(time_integrator)
 add_subdirectory(utils)
+add_subdirectory(output)
```
Comment (zfergus) [inline src/polyfem/CMakeLists.txt:26]: Do we want to use the `io` folder instead of creating a new one?

File: src/polyfem/State.cpp [hunk 5]
```diff
@@ -855,82 +779,14 @@ namespace polyfem
...
-		// std::sort(flipped_elements.begin(), flipped_elements.end());
-		// auto it = std::unique(flipped_elements.begin(), flipped_elements.end());
-		// flipped_elements.resize(std::distance(flipped_elements.begin(), it));
+			stats.count_flipped_elements(*mesh, geom_bases());
 
 		const int prev_bases = n_bases;
 		n_bases += obstacle.n_vertices();
 
-		logger().info("Extracting boundary mesh...");
+		logger().info("Building boundary mesh...");
```
Comment (zfergus) [inline src/polyfem/State.cpp:788]: "Building collision mesh..."?

File: src/polyfem/State.hpp [hunk 6]
```diff
@@ -492,389 +477,74 @@ namespace polyfem
...
 		/// saves the frames in a vector instead of VTU
-		std::vector<SolutionFrame> solution_frames;
-
-		/// extracts the boundary mesh for visualization, called in build_basis
-		void extract_vis_boundary_mesh();
+		std::vector<output::SolutionFrame> solution_frames;
+		/// visualization stuff
+		output::OutGeometryData out_geom;
+		/// runtime statistics
+		output::OutRuntimeData runtime;
```
Comment (zfergus) [inline src/polyfem/State.hpp:503]: `timings` or `running_times`? `runtime` sounds like it is runtime options or something when it really is just the running times.

File: tests/verify_run.cpp [hunk 7]
```diff
@@ -102,7 +102,7 @@ int authenticate_json(std::string json_file, const bool allow_append)
 		return 1;
 	}
 
-	state.compute_mesh_stats();
+	// state.compute_mesh_stats();
```
Comment (zfergus) [inline tests/verify_run.cpp:105]: Is this automatically computed now?

**Review by zfergus — APPROVED** [review]
Looks good! I added some minor comments, feel free to ignore them.

File: tests/verify_run.cpp (hunk 7 above)
Comment (teseoch) [reply tests/verify_run.cpp:105]: it is not necessary for the test, no?

Comment (zfergus) [reply tests/verify_run.cpp:105]: We use 
```
    out["err_l2"] = state.stats.l2_err;
	out["err_h1"] = state.h1_err;
	out["err_h1"] = state.stats.h1_err;
	out["err_h1_semi"] = state.h1_semi_err;
	out["err_h1_semi"] = state.stats.h1_semi_err;
	out["err_linf"] = state.linf_err;
	out["err_linf"] = state.stats.linf_err;
	out["err_linf_grad"] = state.grad_max_err;
	out["err_linf_grad"] = state.stats.grad_max_err;
	out["err_lp"] = state.lp_err;
	out["err_lp"] = state.stats.lp_err;
```
are these always computed?

Comment (teseoch) [reply tests/verify_run.cpp:105]: no, they are computed in `state.compute_errors()`, the other computes the number of pk elements and other mesh stuff

Comment (teseoch) [reply tests/verify_run.cpp:105]: https://github.com/polyfem/polyfem/blob/a11c67e9cfc2dcedf53fc737dbe83f43377ae98e/src/polyfem/io/OutData.cpp#L2387

**Review by zfergus — APPROVED** [review]

**Pull request merged**